Components/CategoryResolver/AutoCategoryResolver.php

```diff
@@ -73,51 +73,47 @@ public function resolve(array $categories)
                 'name' => $node['name'],
                 'parentId' => 1,
             ]);
+            // if connectTree has a Subtree starting with Spanish but MerchantShop has no mainCategory Spanish
+            // the categories below Spanish won't be created
+            if ($mainCategory == null) {
+                continue;
+            }
 
-            $remoteCategories = array_merge($remoteCategories, $this->convertTreeToEntities($node['children'], $mainCategory));
+            $remoteCategories = $this->convertTreeToKeys($node['children'], $mainCategory->getId());
         }
 
         // Collect all, not only leaf categories. Some customers use them to assign products.
         // Do not fetch them from database by name as before.
         // it is possible to have more than one subcategory "Boots" - CON-4589
         return array_map(function ($category) {
-            return $category['model'];
+            return $category['categoryKey'];
         }, $remoteCategories);
     }
 
     /**
-     * Loop categories tree recursive and
-     * create same structure with entities
+     * Loop through category tree and fetch ids
      *
      * @param array $node
-     * @param null Category $parent
+     * @param string $parentId
+     * @param bool $returnOnlyLeafs
      * @param array $categories
      * @return array
      */
-    public function convertTreeToEntities(array $node, Category $parent = null, $categories = [])
+    public function convertTreeToKeys(array $node, $parentId, $returnOnlyLeafs = true, $categories = [])
```

Comment:
If every public function should be defined in CategoryResolver abstract class as abstract function or implemented there and available for every different implementation of CategoryResolver, otherwise it must be private, but we need it in ImportService.

Reply:
i just move it to the abstract class because there is just this one implementation